File: examples/java/src/main/java/io/superdurable/dex/patterns/parallelsubflows/AdvancedShortLiveParentFlow.java
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 */

package io.superdurable.dex.patterns.parallelsubflows;

import io.superdurable.dex.Attribute;
import io.superdurable.dex.AttributeLock;
import io.superdurable.dex.Channel;
import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.PersistenceSchema;
import io.superdurable.dex.RPC;
import io.superdurable.dex.RPCResult;
import io.superdurable.dex.Step;
import io.superdurable.dex.StepDecision;
import io.superdurable.dex.StepList;
import io.superdurable.dex.StepMovement;
import io.superdurable.dex.StepOptions;
import io.superdurable.dex.SubFlow;
import io.superdurable.dex.Wait;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class AdvancedShortLiveParentFlow implements Flow<ParentInput> {
    public static final int DEFAULT_CONCURRENCY = 10;
    public static final int MAX_BUFFERED_REQUESTS = 100;

    public final Channel<String> requestChannel = Channel.define("RequestChannel", String.class);
    public final Attribute<Integer> currSubFlowNum = Attribute.define("CurrSubFlowNum", Integer.class);

    private final InitStep initStep = new InitStep();
    private final HandleRequestStep handleRequestStep = new HandleRequestStep();
    private final HandleSubFlowStep handleSubFlowStep;

    public AdvancedShortLiveParentFlow(final ExampleSubFlow exampleSubFlow) {
        handleSubFlowStep = new HandleSubFlowStep(exampleSubFlow);
    }

    @Override
    public StepList<ParentInput> getSteps() {
        return StepList.startStep(initStep).otherSteps(handleRequestStep, handleSubFlowStep);
    }

    @Override
    public PersistenceSchema getPersistenceSchema() {
        return PersistenceSchema.of(requestChannel, currSubFlowNum);
    }

    @RPC
    public RPCResult<Boolean> sendRequest(final Context context, final String request) {
        if (requestChannel.size(context) >= MAX_BUFFERED_REQUESTS) {
            return RPCResult.of(false);
        }
        requestChannel.publish(context, request);
        return RPCResult.of(true);
    }

    final class InitStep implements Step<ParentInput> {
        @Override
        public Class<ParentInput> getInputType() {
            return ParentInput.class;
        }

        @Override
        public StepDecision execute(final Context context, final ParentInput input) {
            for (final String request : input.requests) {
                requestChannel.publish(context, request);
            }
            currSubFlowNum.set(context, 0);
            final int concurrency = input.concurrency > 0 ? input.concurrency : DEFAULT_CONCURRENCY;
            final StepMovement<?>[] movements = new StepMovement<?>[concurrency];
            for (int index = 0; index < concurrency; index++) {
                movements[index] = StepMovement.of(HandleRequestStep.class, null);
            }
            return StepDecision.goToMany(movements);
        }
    }

    final class HandleRequestStep implements Step<Void> {
        @Override
        public Class<Void> getInputType() {
            return Void.class;
        }

        @Override
        public StepOptions getStepOptions() {
            return StepOptions.newBuilder().addExecuteLock(AttributeLock.of(currSubFlowNum)).build();
        }

        @Override
        public Wait waitFor(final Context context, final Void input) {
            return Wait.until(requestChannel.forOne());
        }

        @Override
        public StepDecision execute(final Context context, final Void input) {
            final List<String> requests = requestChannel.getConditionResults(context);
            currSubFlowNum.set(context, currSubFlowNum.get(context) + 1);
            return StepDecision.goTo(HandleSubFlowStep.class, requests.get(0));
        }
    }

    final class HandleSubFlowStep implements Step<String> {
        private final ExampleSubFlow exampleSubFlow;

        HandleSubFlowStep(final ExampleSubFlow exampleSubFlow) {
            this.exampleSubFlow = exampleSubFlow;
        }

        @Override
        public Class<String> getInputType() {
            return String.class;
        }

        @Override
        public StepOptions getStepOptions() {
            return StepOptions.newBuilder().addExecuteLock(AttributeLock.of(currSubFlowNum)).build();
        }

        @Override
        public Wait waitFor(final Context context, final String request) {
            return Wait.until(SubFlow.run(ExampleSubFlow.class, request));
        }

        @Override
        public StepDecision execute(final Context context, final String request) {
            final int current = currSubFlowNum.get(context) - 1;
            currSubFlowNum.set(context, current);
            if (current == 0) {
                return StepDecision.forceCompleteIfChannelsEmpty(
                        null,
                        StepMovement.of(HandleRequestStep.class, null),
                        requestChannel);
            }
            return StepDecision.goTo(HandleRequestStep.class, null);
        }
    }
}

File: examples/go/patterns/parallel-subflows/advanced_short_live_parent_flow.go
// Copyright (c) 2022-2026 Super Durable, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

package parallelsubflows

import "github.com/superdurable/dex/sdk-go/dex"

var (
	RequestChannel = dex.DefineChannel[string]("RequestChannel")
	CurrSubFlowNum = dex.DefineAttribute[int]("CurrSubFlowNum")
)

type AdvancedShortLiveParentFlow struct {
	dex.FlowDefaults
	exampleFlow *ExampleSubFlow
}

func NewAdvancedShortLiveParentFlow(exampleFlow *ExampleSubFlow) *AdvancedShortLiveParentFlow {
	return &AdvancedShortLiveParentFlow{exampleFlow: exampleFlow}
}

func (flow *AdvancedShortLiveParentFlow) GetSteps() []dex.StepDef {
	return []dex.StepDef{
		dex.DefineStartStep(shortLiveInitStep{}),
		dex.DefineStep(shortLiveHandleRequestStep{}),
		dex.DefineStep(shortLiveHandleSubFlowStep{exampleFlow: flow.exampleFlow}),
	}
}

func (*AdvancedShortLiveParentFlow) GetPersistenceSchema() dex.PersistenceSchema {
	return dex.PersistenceSchema{
		Attributes: []dex.AttributeDef{CurrSubFlowNum},
		Channels:   []dex.ChannelDef{RequestChannel},
	}
}

func (*AdvancedShortLiveParentFlow) SendRequest(ctx dex.Context, request string) (*dex.RPCResult[bool], error) {
	if RequestChannel.Size(ctx) >= MaxBufferedRequests {
		return &dex.RPCResult[bool]{Output: false}, nil
	}
	if err := RequestChannel.Publish(ctx, request); err != nil {
		return nil, err
	}
	return &dex.RPCResult[bool]{Output: true}, nil
}

type shortLiveInitStep struct {
	dex.StepDefaultsNoWaitFor[ParentInput]
}

func (shortLiveInitStep) GetStepType() string { return "InitStep" }

func (shortLiveInitStep) Execute(ctx dex.Context, input ParentInput) (*dex.StepDecision, error) {
	for _, request := range input.Requests {
		if err := RequestChannel.Publish(ctx, request); err != nil {
			return nil, err
		}
	}
	if err := CurrSubFlowNum.Set(ctx, 0); err != nil {
		return nil, err
	}
	count := concurrency(input.Concurrency)
	movements := make([]dex.StepMovement, 0, count)
	for index := 0; index < count; index++ {
		movements = append(movements, dex.MovementOf(shortLiveHandleRequestStep{}, nil))
	}
	return dex.GoToMany(movements...), nil
}

type shortLiveHandleRequestStep struct{ dex.StepDefaults }

func (shortLiveHandleRequestStep) GetStepType() string { return "HandleRequestStep" }

func (shortLiveHandleRequestStep) GetStepOptions() *dex.StepOptions {
	return &dex.StepOptions{ExecuteLockAttributes: []dex.AttributeLock{dex.LockAttribute(CurrSubFlowNum)}}
}

func (shortLiveHandleRequestStep) WaitFor(_ dex.Context, _ dex.None) (*dex.Wait, error) {
	return dex.Until(RequestChannel.ForOne()), nil
}

func (shortLiveHandleRequestStep) Execute(ctx dex.Context, _ dex.None) (*dex.StepDecision, error) {
	requests, err := RequestChannel.GetConditionResults(ctx)
	if err != nil {
		return nil, err
	}
	current, err := CurrSubFlowNum.Get(ctx)
	if err != nil {
		return nil, err
	}
	if err := CurrSubFlowNum.Set(ctx, current+1); err != nil {
		return nil, err
	}
	return dex.GoTo(shortLiveHandleSubFlowStep{}, requests[0]), nil
}

type shortLiveHandleSubFlowStep struct {
	dex.StepDefaults
	exampleFlow *ExampleSubFlow
}

func (shortLiveHandleSubFlowStep) GetStepType() string { return "HandleSubFlowStep" }

func (shortLiveHandleSubFlowStep) GetStepOptions() *dex.StepOptions {
	return &dex.StepOptions{ExecuteLockAttributes: []dex.AttributeLock{dex.LockAttribute(CurrSubFlowNum)}}
}

func (step shortLiveHandleSubFlowStep) WaitFor(_ dex.Context, request string) (*dex.Wait, error) {
	return dex.Until(dex.SubFlow(step.exampleFlow, request)), nil
}

func (shortLiveHandleSubFlowStep) Execute(ctx dex.Context, _ string) (*dex.StepDecision, error) {
	current, err := CurrSubFlowNum.Get(ctx)
	if err != nil {
		return nil, err
	}
	current--
	if err := CurrSubFlowNum.Set(ctx, current); err != nil {
		return nil, err
	}
	if current == 0 {
		return dex.ForceCompleteIfChannelsEmpty(
			nil,
			[]dex.ChannelDef{RequestChannel},
			dex.MovementOf(shortLiveHandleRequestStep{}, nil),
		), nil
	}
	return dex.GoTo(shortLiveHandleRequestStep{}, nil), nil
}

var (
	_ dex.Flow              = (*AdvancedShortLiveParentFlow)(nil)
	_ dex.RPC[string, bool] = (*AdvancedShortLiveParentFlow)(nil).SendRequest
)
